internal/querycoordv2/balance/score_based_balancer.go

```diff
@@ -0,0 +1,387 @@
+	return &ScoreBasedBalancer{
+		RowCountBasedBalancer:           NewRowCountBasedBalancer(scheduler, nodeManager, dist, meta, targetMgr),
+		balancedCollectionsCurrentRound: typeutil.NewUniqueSet(),
+	}
+}
+
+// TODO assign channel need to think of global channels
+func (b *ScoreBasedBalancer) AssignSegment(collectionID int64, segments []*meta.Segment, nodes []int64) []SegmentAssignPlan {
+	nodeItems := b.convertToNodeItems(collectionID, nodes)
+	if len(nodeItems) == 0 {
+		return nil
+	}
+	queue := newPriorityQueue()
+	for _, item := range nodeItems {
+		queue.push(item)
+	}
+
+	sort.Slice(segments, func(i, j int) bool {
+		return segments[i].GetNumOfRows() > segments[j].GetNumOfRows()
+	})
+
+	plans := make([]SegmentAssignPlan, 0, len(segments))
+	for _, s := range segments {
+		// pick the node with the least row count and allocate to it.
+		ni := queue.pop().(*nodeItem)
+		plan := SegmentAssignPlan{
+			From:    -1,
+			To:      ni.nodeID,
+			Weight:  GetWeight(1),
+			Segment: s,
+		}
+		plans = append(plans, plan)
+		// change node's priority and push back, should count for both collection factor and local factor
+		p := ni.getPriority()
+		ni.setPriority(p + int(s.GetNumOfRows()) +
+			int(float64(s.GetNumOfRows())*params.Params.QueryCoordCfg.GlobalRowCountFactor.GetAsFloat()))
+		queue.push(ni)
+	}
+	return plans
+}
+
+func (b *ScoreBasedBalancer) convertToNodeItems(collectionID int64, nodeIDs []int64) []*nodeItem {
+	ret := make([]*nodeItem, 0, len(nodeIDs))
+	for _, nodeInfo := range b.getNodes(nodeIDs) {
+		node := nodeInfo.ID()
+		priority := b.calculatePriority(collectionID, node)
+		nodeItem := newNodeItem(priority, node)
+		ret = append(ret, &nodeItem)
+	}
+	return ret
+}
+
+func (b *ScoreBasedBalancer) calculatePriority(collectionID, nodeID int64) int {
+	globalSegments := b.dist.SegmentDistManager.GetByNode(nodeID)
+	rowCount := 0
+	for _, s := range globalSegments {
+		rowCount += int(s.GetNumOfRows())
+	}
+
+	collectionSegments := b.dist.SegmentDistManager.GetByCollectionAndNode(collectionID, nodeID)
+	collectionRowCount := 0
+	for _, s := range collectionSegments {
+		collectionRowCount += int(s.GetNumOfRows())
+	}
+	return collectionRowCount + int(float64(rowCount)*
+		params.Params.QueryCoordCfg.GlobalRowCountFactor.GetAsFloat())
+}
+
+func (b *ScoreBasedBalancer) Balance() ([]SegmentAssignPlan, []ChannelAssignPlan) {
+	ids := b.meta.CollectionManager.GetAll()
+
+	// loading collection should skip balance
+	loadedCollections := lo.Filter(ids, func(cid int64, _ int) bool {
+		return b.meta.GetStatus(cid) == querypb.LoadStatus_Loaded
+	})
+
+	sort.Slice(loadedCollections, func(i, j int) bool {
+		return loadedCollections[i] < loadedCollections[j]
+	})
+
+	segmentPlans, channelPlans := make([]SegmentAssignPlan, 0), make([]ChannelAssignPlan, 0)
+	hasUnBalancedCollections := false
+	for _, cid := range loadedCollections {
+		if b.balancedCollectionsCurrentRound.Contain(cid) {
```

Comment:
under what case will we see balancedCollectionsCurrentRound with same collection ID?